Add ItemCoreOrder config to EDS.ini and use it to order EDS item fields

alexklbuckley opened the pull request:
This enhancement enables users to define in the EDS.ini file the labels of EDS fields returned from the EDS API and then set a order value for how they should be displayed in the striped table on the Eds Record page (core.phtml). 

For example:
```
[ItemCoreOrder]
Title =  0
Authors = 1
Source = 2
Abstract = 3
Subject Terms = 4
```

This would mean the Title field is displayed first and the Subject terms displayed last. 
If some, but not all, field labels are defined under ItemCoreOrder then they are assigned a default order.

If no ItemCoreOrder values are defined then they are displayed in the order the EDS API returns them in.

alexklbuckley committed:
Add ItemCoreOrder config to EDS.ini and use it to order how item fields
are displayed in the EDS core.phtml page

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -335,6 +335,17 @@ AU = None
+; This section can be used to order items based on their labels.
+[ItemCoreOrder]
+; Note the label of the item field from the EDS API first then the relative order it should be shown in
+;Title         = 0
```
The .ini parser is more sensitive about which characters are allowed in a key than in a value. Might it be a good idea to reverse this configuration so the numbers are the keys, just in case there is ever a field label containing a character that is illegal in a key name?

alexklbuckley replied:
Great idea Demian, thank you! 
I've pushed that change in a follow-up commit on this PR. I also set the example first key order value to 1 not 0, so that value is not ignored by later logic and for usability :)

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
```diff
@@ -258,22 +258,41 @@ public function getItems(
+            $orig_items   = $this->fields['Items'];
+            $pos          = '';
+            $defaultOrder = count($orig_items);
+            $ItemCoreOrderConfig = isset($this->recordConfig->ItemCoreOrder)
```
Style: variables should start with lowercase letters, so I'd recommend changing this to `$itemCoreOrderConfig`.

alexklbuckley replied:
Done in my follow-up commit :)

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
```diff
@@ -258,22 +258,41 @@ public function getItems(
+	    $count = count($ItemCoreOrderConfig, COUNT_RECURSIVE);
+            if ($count > 0) {
+                for ($i = 0; $i < count($orig_items); $i++) {
+                    $label_name = $orig_items[$i]["Label"];
```
Style: we prefer single quotes when possible -- if you run `vendor/bin/phing php-cs-fixer` from your $VUFIND_HOME directory, it should automatically fix this for you.

alexklbuckley replied:
Done in my follow-up commit

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -335,6 +335,17 @@ AU = None
+;Abstract      = 3
+;Subject Terms = 4
+
+
```
Style: I don't think we need three blank lines between sections; one is probably enough.

alexklbuckley replied:
Fixed up in my follow-up commit

demiankatz reviewed (changes requested):
Thanks for getting this started, @alexklbuckley!

I've left a few comments, mostly about project style things that you might find helpful to know. The other thing you may find helpful along these lines is that running `composer run qa` will perform all the same tests as GitHub continuous integration, and it can be useful for catching certain common issues. (Of course, it's also fine to just push things up to the PR and let GitHub take care of it... but since this is your first PR, the checks don't run automatically yet, so you won't get feedback until somebody else in the project approves the run).

Beyond that, my biggest concern is that I see code to set a 'Pos' value, but I don't see any code that actually reorders things based on that value. Am I overlooking something, or does more code need to be added to make this fully functional? (I confess that I haven't actually tried running the code yet, but I'm happy to do some hands-on testing when you're ready).

alexklbuckley committed:
Implement community feedback for ordering item fields in the EDS record
page

- Reverse the ItemCoreOrder key and value pairs so the order number is the
key.

- Correct variable name and use single quotes in line with styling
guidelines.

- Start the ItemCoreOrder array keys at 1 not 0, for usability and so
  the first ordered field is not ignored.

- Check if the EDS API item field name is a value in the ItemCoreOrder
associative array.

alexklbuckley commented:
Thanks so much for taking a look Demian! You're absolutely right I did miss adding the code that does the reordering based on the 'Pos' value, I have added that into my follow-up commit. 

I am writing this feature for a VuFind 9.0.3 instance, so I haven't tested running this code on the dev branch, only on the VuFind 9.0.3 branch, so I hope this is now ready for testing, but happy to fix anything else up that you notice.

Thanks again!

demiankatz committed:
Merge branch 'dev' into eds-itemcoreorder

demiankatz committed:
php-cs-fixer

demiankatz committed:
Code simplificaiton/fixes.

demiankatz committed:
More logical checks.

demiankatz committed:
Tweak comments.

demiankatz commented:
Thanks, @alexklbuckley! I spent a little bit of time testing this today. It wasn't quite working for me, so I did some minor restructuring to get it to work (the main problem was that $orig_items was not a reference, so making changes to it was having no effect on the displayed order in the code as written). I've restructured the code to avoid the reference problem, plus I've made some minor stylistic changes (using $camelCase instead of $snake_case for consistency with the prevailing style, taking advantage of some newer PHP 8 language features, and doing a bit of refactoring to reduce redundancy here and there). I also added logic to skip the sort except when it's actually needed, to improve performance.

Do you mind reviewing/testing my latest version and letting me know if it looks okay and works correctly for you?

demiankatz committed:
php-cs-fixer

alexklbuckley commented:
Thanks so much for testing and making those changes Demian!

I've tested the code changes locally and it works well for me. Reviewing the code itself , the changes you have made make perfect sense, especially around the $orig_items.

The only question I have is, am I right in thinking that lines [271](https://github.com/vufind-org/vufind/pull/3284/commits/afc17493023e66efe69f2c79365832962496b5f7#diff-537d9dd65e722a091d5c96a3014308932a1c377205e0138a0736b1ff7fc90cc8R271) and [272](https://github.com/vufind-org/vufind/pull/3284/commits/afc17493023e66efe69f2c79365832962496b5f7#diff-537d9dd65e722a091d5c96a3014308932a1c377205e0138a0736b1ff7fc90cc8R272) are doing the following:

- If the current item field label is not defined in the configuration 
Then set the 'pos' value to nextPos (total number of EDS field items + 1)
- Else set the 'pos' value to whatever it's defined as in the config.

Is that correct?

demiankatz commented:
> Thanks so much for testing and making those changes Demian!

Happy to help!

> I've tested the code changes locally and it works well for me. Reviewing the code itself , the changes you have made make perfect sense, especially around the $orig_items.

Thanks!

> The only question I have is, am I right in thinking that lines [271](https://github.com/vufind-org/vufind/pull/3284/commits/afc17493023e66efe69f2c79365832962496b5f7#diff-537d9dd65e722a091d5c96a3014308932a1c377205e0138a0736b1ff7fc90cc8R271) and [272](https://github.com/vufind-org/vufind/pull/3284/commits/afc17493023e66efe69f2c79365832962496b5f7#diff-537d9dd65e722a091d5c96a3014308932a1c377205e0138a0736b1ff7fc90cc8R272) are doing the following:
> 
>     * If the current item field label is not defined in the configuration
>       Then set the 'pos' value to nextPos (total number of EDS field items + 1)
> 
>     * Else set the 'pos' value to whatever it's defined as in the config.
> 
> 
> Is that correct?

Yes, that's correct. I'm using the ternary operator to avoid an if/else statement and duplicate assignment logic, but it's logically equivalent to:

```
                    $configuredPos = array_search($label, $itemCoreOrderConfig);
                    if ($configuredPos === false) {
                        $origItems[$key]['Pos'] = $nextPos++;
                    } else {
                        $origItems[$key]['Pos'] = $configuredPos;
                    }
```

Let me know if that's okay with you, and if it is, I'll see if anyone else wants to review this, and get it merged once there's consensus.

demiankatz commented:
I suspect this code might be of some interest to @maccabeelevine, so requesting a review...

alexklbuckley commented:
Thanks for the clarification about the ternary operator, good to know.  Definitely okay with it and happy to see if others want to review and get it merged once there's consensus.

Also, thank you for your notes during this PR it's been really useful for me getting more understanding as I start doing more VuFind development work :)

demiankatz commented:
> Thanks for the clarification about the ternary operator, good to know. Definitely okay with it and happy to see if others want to review and get it merged once there's consensus.

Great, I'll give this a couple of days to collect any further feedback and will likely merge next week if there are no objections/suggestions.

> Also, thank you for your notes during this PR it's been really useful for me getting more understanding as I start doing more VuFind development work :)

Of course, happy to help, and please feel free to ask questions at any time if there's anything else I can help with. I'm always grateful to have new contributors, so I do what I can to make the onboarding as painless as possible. ;-)

maccabeelevine commented on config/vufind/EDS.ini:
```diff
@@ -335,6 +335,15 @@ AU = None
 ; This repeatable setting can be used to filter out values using group codes:
 ;excludeGroup[] = 'URL'
 
+; This section can be used to order items based on their labels.
```
It may be worth mentioning that any items defined here will appear before any items not defined here.  I.e. if the only value supplied is `5 = Title`, the title will still be listed first.

maccabeelevine replied:
Actually I see my statement above was wrong, it's more complicated than that.  If the only value supplied is `500 = Title`, then title will actually appear last.  Since for 'n' original items the defaults are ordered between 'n' and '2n'.  So perhaps easier just to tell people they should order consecutively starting at 1, so all of the configured items appear before non-configured ones, and not to tell them what happens if they do the numbers differently. :)

alexklbuckley replied:
Yeah agreed! I'll write a patch doing that

alexklbuckley replied:
I've pushed this change in the follow-up patchset :)

maccabeelevine commented on config/vufind/EDS.ini:
```diff
@@ -335,6 +335,15 @@ AU = None
 ;excludeGroup[] = 'URL'
 
+; This section can be used to order items based on their labels.
+[ItemCoreOrder]
```
I was not sure what the word "Core" meant here, but I see another section of EDS.ini says 

> ; Similar to [ItemGlobalFilter] but applied only for record view.
> [ItemCoreFilter]

So I think the word is misleading then, since the ItemCoreOrder as written applies to both the results page and the record view.  Would ItemGlobalOrder be correct?

alexklbuckley replied:
I've pushed this change to ItemGlobalOrder in my follow-up patchset, and switched EDS.php to use that new config name and have corrected variable names.

maccabeelevine reviewed (changes requested):
This is a nice feature addition!  Just two suggestions on naming / wording.

alexklbuckley commented:
Thanks @maccabeelevine for raising those good points!

alexklbuckley committed:
Wording changes to the EDS.ini

- Clarify that EDS item fields should have an order value starting at 1

- Rename ItemCoreOrder to ItemGlobalOrder - given that this setting
  affects both the results page and record view

alexklbuckley committed:
Switch EDS.php to use ItemGlobalOrder as the configuration setting name

maccabeelevine reviewed (approved):
LGTM!

demiankatz committed:
Switch to pre-increment.

demiankatz committed:
Fix $nextPos determination.

demiankatz commented:
I gave this one last review and found and fixed one last minor issue: the calculation we were using to determine how to position rows after the configured order was incorrect. We were basing it on number of data rows in the current item, but that doesn't really reliably tell us anything -- it should actually be based on the maximum configured position to ensure that non-configured rows always appear after configured rows. The last two commits work together to fix this. I think we're good now, so I'm going to merge this!

alexklbuckley commented:
Thanks so much Demian and @maccabeelevine !!!
